src/preamble.js

```diff
@@ -2541,15 +2541,15 @@ function integrateWasmJS() {
         abort('bad method: ' + curr);
       }
     }
-
-    if (!exports) throw 'no binaryen method succeeded. consider enabling more options, like interpreting, if you want that: https://github.com/kripken/emscripten/wiki/WebAssembly#binaryen-methods';
 #endif
 #endif
 
 #if RUNTIME_LOGGING
     Module['printErr']('binaryen method succeeded.');
 #endif
 
+    if (!exports) abort('no binaryen method succeeded. consider enabling more options, like interpreting, if you want that: https://github.com/kripken/emscripten/wiki/WebAssembly#binaryen-methods');
```

Comment:
thanks, this does look like it fixes the problem. but please move it before that logging of "succeeded", so that when it fails it won't contradict itself in the log ;)

Reply:
also we should add a test for this, among the `other.test_binaryen*` ones probably. maybe something like delete the wasm file, then see that it fails with that error message.

Reply:
lol, oops. Thanks, just pushed a fix.

Reply:
Well, the case where this was causing a problem was when the environment didn't support WebAssembly. Just failing to compile wasm aborted as expected.

Reply:
Hmm, how did it abort as expected before, if that abort wasn't in the right place? Testing now it seems to print `no native wasm support detected` but it doesn't abort - it keeps running, then hits an error when it tries to use wasm code (malloc). Maybe we're testing different options.

Anyhow, maybe we don't need a test here, but I'd like to fully understand what changes here.

Reply:
Ah right, sorry, probably should've explained that. Without this PR, there are a few places in `doNativeWasm` where it exits early with a `return false`, such as when there's no native wasm support, e.g.

https://github.com/kripken/emscripten/blob/5d18b58b00c45bced5f4cea4d233cf86eb1af2c3/src/preamble.js#L2272

In some cases, such as when wasm compilation fails (and I think when the wasm binary isn't present at all, but not 100% sure offhand), it'll abort on line 2335; but the general failure case just keeps running as you experienced.